src/Sarif/ExtensionMethods.cs

```diff
@@ -514,6 +514,45 @@
             return propertyBag;
         }
 
+        public static IEnumerable<T> MergeWithList<T>(
+            this IEnumerable<T> list,
+            IEnumerable<T> anotherList, 
+            IEqualityComparer<T> equalityComparer)
+        {
+            if (list == null)
+            {
+                throw new ArgumentNullException(nameof(list));
+            }
+
+            if (anotherList == null)
+            {
+                throw new ArgumentNullException(nameof(anotherList));
+            }
+
+            if (equalityComparer == null)
+            {
+                throw new ArgumentNullException(nameof(equalityComparer));
+            }
+
+            var elementSet = new HashSet<T>(equalityComparer);
+
+            foreach (T element in list)
+            {
+                if (elementSet.Add(element))
+                {
+                    yield return element;
+                }
+            }
```

Comment:
Right, in this case the suggestion doesn't make sense because it has to iterate all elements and yield return the element, change to `Where` call make it difficult to understand.

In some cases the `Where` may be useful to make a complicated filtering condition to be a short, easy to read standard query, with little performance trade-off.

I think we can keep the codeQL rule, and check specific case if its applies. :grinning: